media-source/mediasource-duration-buffered.html

```diff
@@ -0,0 +1,100 @@
+            sourceBuffer.duration = segmentInfo.duration;
+            var midSegment = segmentInfo.media[segmentInfo.media.length-1];
+            var newDuration = midSegment.timecode / 2.0;
+            var midData = MediaSourceUtil.extractSegmentData(mediaData, midSegment);
+            test.expectEvent(sourceBuffer, "updateend");
+            sourceBuffer.appendBuffer(midData);
+            test.waitForExpectedEvents(function()
+            {
+                assert_equals(sourceBuffer.buffered.length, 1);
+                assert_greater_than(sourceBuffer.buffered.start(0), newDuration);
+                assert_throws("InvalidStateError", function ()
+                {
+                    mediaSource.duration = newDuration;
+                });
+                test.done();
+            });
+        });
+    }, "MediaSource.duration: Truncating the duration throws an InvalidStateError exception when new duration is less than the highest starting time of buffered coded frames");
+
+
+    mediasource_test(function(test, mediaElement, mediaSource)
+    {
+        mediaElement.addEventListener("error", test.unreached_func("Unexpected event 'error'"));
+        MediaSourceUtil.fetchManifestAndData(test, manifestFilenameAudio, function (typeAudio, dataAudio)
+        {
+            MediaSourceUtil.fetchManifestAndData(test, manifestFilenameVideo, function (typeVideo, dataVideo)
+            {
+                var sourceBufferAudio = mediaSource.addSourceBuffer(typeAudio);
+                var sourceBufferVideo = mediaSource.addSourceBuffer(typeVideo);
+                var newDuration = 1.2;
+
+                sourceBufferAudio.appendWindowEnd = 2.0;
+                sourceBufferAudio.appendWindowStart = newDuration / 2.0;
+                sourceBufferAudio.appendBuffer(dataAudio);
+
+                sourceBufferVideo.appendWindowEnd = 2.0;
+                sourceBufferVideo.appendWindowStart = newDuration * 1.3;
+                sourceBufferVideo.appendBuffer(dataVideo);
+
+                test.expectEvent(sourceBufferAudio, "updateend");
+                test.expectEvent(sourceBufferVideo, "updateend");
+                test.waitForExpectedEvents(function()
+                {
+                    assert_equals(sourceBufferAudio.buffered.length, 1);
+                    assert_equals(sourceBufferVideo.buffered.length, 1);
+                    assert_less_than(sourceBufferAudio.buffered.start(0), newDuration);
+                    assert_greater_than(sourceBufferVideo.buffered.start(0), newDuration);
+                    assert_throws("InvalidStateError", function ()
+                    {
+                        mediaSource.duration = newDuration;
+                    });
+                    test.done();
+                });
+            });
+        });
+    }, "MediaSource.duration: Truncating the duration throws an InvalidStateError exception when new duration is less than the highest starting time for at least some buffered coded frames");
```

Comment:
This subtest is a little unique versus the current mediasource-duration.html "Test setting the duration to less than the highest starting presentation timestamp will throw", since it uses timestampOffset to directly manipulate that audio and video buffered range vs the newDuration, thereby catching broken implementations which might allow truncating to highest HTMLME.buffered end time (disregarding the underlying track buffer ranges highest presentation start time). I'll prepare an update to the current mediasource-duration.html which includes this deeper subtest.